File: src/runtime/jsonrpc/messages.ts
// Hand-rolled JSON-RPC 2.0 message types shared by the runtime socket server and
// any TS client. Kept dependency-free and mirrored 1:1 by the Go package in
// `tui/internal/jsonrpc`. See docs/go-tui-migration-plan.md §4 / §4.1.

export type JsonRpcId = number | string

export type JsonRpcRequest = {
  jsonrpc: "2.0"
  id: JsonRpcId
  method: string
  params?: unknown
}

export type JsonRpcNotification = {
  jsonrpc: "2.0"
  method: string
  params?: unknown
}

export type JsonRpcSuccess = {
  jsonrpc: "2.0"
  id: JsonRpcId
  result: unknown
}

export type JsonRpcFailure = {
  jsonrpc: "2.0"
  id: JsonRpcId
  error: JsonRpcErrorObject
}

export type JsonRpcResponse = JsonRpcSuccess | JsonRpcFailure

export type JsonRpcMessage =
  | JsonRpcRequest
  | JsonRpcNotification
  | JsonRpcResponse

export type JsonRpcErrorObject = {
  code: number
  message: string
  data?: unknown
}

// Reserved range (-32700..-32600) is for protocol-level errors; application
// errors use -32000 and below, discriminated by `data.kind` (plan §4).
export const JsonRpcErrorCode = {
  ParseError: -32700,
  InvalidRequest: -32600,
  MethodNotFound: -32601,
  InvalidParams: -32602,
  InternalError: -32603,
} as const

// Throw this from a handler to fail a request with a specific code/data. Any
// other thrown value is reported as InternalError.
export class RpcError extends Error {
  readonly code: number
  readonly data?: unknown

  constructor(code: number, message: string, data?: unknown) {
    super(message)
    this.name = "RpcError"
    this.code = code
    this.data = data
  }

  toErrorObject(): JsonRpcErrorObject {
    return this.data === undefined
      ? { code: this.code, message: this.message }
      : { code: this.code, message: this.message, data: this.data }
  }
}

// A response carries no `method`; a request/notification always does.
export const isResponse = (
  message: JsonRpcMessage,
): message is JsonRpcResponse => !("method" in message)

export const isFailure = (
  response: JsonRpcResponse,
): response is JsonRpcFailure => "error" in response

File: tui/internal/jsonrpc/messages.go
// Package jsonrpc is a hand-rolled, dependency-free JSON-RPC 2.0 endpoint that
// mirrors the TypeScript implementation in src/runtime/jsonrpc. It speaks the
// same LSP-style Content-Length framing so the Go TUI and the TS runtime can
// talk over a Unix-domain socket. See docs/go-tui-migration-plan.md §4 / §4.1.
package jsonrpc

import "encoding/json"

// Message is a single JSON-RPC frame. The kind is inferred from which fields are
// present, matching the TS side:
//
//	request:      Method set, ID set
//	notification: Method set, ID nil
//	response:     Method empty, ID set, Result or Error set
//
// ID is a raw value so it round-trips numbers or strings unchanged.
type Message struct {
	JSONRPC string           `json:"jsonrpc"`
	ID      *json.RawMessage `json:"id,omitempty"`
	Method  string           `json:"method,omitempty"`
	Params  json.RawMessage  `json:"params,omitempty"`
	Result  json.RawMessage  `json:"result,omitempty"`
	Error   *Error           `json:"error,omitempty"`
}

// Error is the JSON-RPC error object. It implements the error interface so it
// can be returned directly from Request.
type Error struct {
	Code    int             `json:"code"`
	Message string          `json:"message"`
	Data    json.RawMessage `json:"data,omitempty"`
}

func (e *Error) Error() string { return e.Message }

// Reserved range (-32700..-32600) is for protocol-level errors; application
// errors use -32000 and below, discriminated by a `kind` field in Data.
const (
	ParseError     = -32700
	InvalidRequest = -32600
	MethodNotFound = -32601
	InvalidParams  = -32602
	InternalError  = -32603
)

// isResponse reports whether the message is a response (no method) rather than a
// request or notification.
func (m *Message) isResponse() bool { return m.Method == "" }
